extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/shuffle/output/FutureReadableFrameChannel.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.msq.shuffle.output;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import org.apache.druid.common.guava.FutureUtils;
+import org.apache.druid.frame.Frame;
+import org.apache.druid.frame.channel.ReadableFrameChannel;
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import java.util.NoSuchElementException;
+
+public class FutureReadableFrameChannel implements ReadableFrameChannel
+{
+  private static final Logger log = new Logger(FutureReadableFrameChannel.class);
+
+  private final ListenableFuture<ReadableFrameChannel> channelFuture;
+  private ReadableFrameChannel channel;
+
+  public FutureReadableFrameChannel(final ListenableFuture<ReadableFrameChannel> channelFuture)
+  {
+    this.channelFuture = channelFuture;
+  }
+
+  @Override
+  public boolean isFinished()
+  {
+    if (populateChannel()) {
+      return channel.isFinished();
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public boolean canRead()
+  {
+    if (populateChannel()) {
+      return channel.canRead();
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public Frame read()
+  {
+    if (populateChannel()) {
+      return channel.read();
+    } else {
+      throw new NoSuchElementException();
+    }
+  }
+
+  @Override
+  public ListenableFuture<?> readabilityFuture()
+  {
+    if (populateChannel()) {
+      return channel.readabilityFuture();
+    } else {
+      return FutureUtils.transformAsync(channelFuture, ignored -> readabilityFuture());
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    if (populateChannel()) {
+      channel.close();
+    } else {
+      channelFuture.cancel(true);
+      channelFuture.addListener(
```

Comment:
Quick question - Why are we attaching a listener after canceling the future? Is it a defensive check for closing the channel if the `cancel` didn't register "in time"? Also given that this is happening on a single thread would we ever encounter such case when the addListener could get triggered?

Reply:
Yeah, it's in case of a race where `channelFuture` resolved after `populateChannel()` but before the call to `cancel`. In that case the call to `cancel` would have no effect, and the listener would fire to clean up the channel.

I don't think this needs to be a listener, really, it could just be in-line. I edited the code to make this logic in-line, and added a comment.